qualcommax: fix PCIe DTC warnings

robimarko opened the pull request:
Currently, DTC is throwing PCIe node warnings:
```
arch/arm64/boot/dts/qcom/ipq8072-wpq873.dts:473.13-478.4: Warning (pci_device_reg): /soc@0/pci@10000000/bridge@1,0: PCI unit address format error, expected "0,0"
arch/arm64/boot/dts/qcom/ipq8072-haze.dts:303.13-318.4: Warning (pci_device_reg): /soc@0/pci@10000000/bridge@1,0: PCI unit address format error, expected "0,0"
arch/arm64/boot/dts/qcom/ipq8072-ax9000.dts:566.13-582.4: Warning (pci_device_reg): /soc@0/pci@10000000/bridge@1,0: PCI unit address format error, expected "0,0"
```

As we are going to have to update DTS for 6.12 since upstream added PCIe bridge nodes we might as well do it now and fix these warnings along the way.

So:
1. Backport IPQ807x and IPQ60xx PCIe bridge node commits
2. Remove our downstream PCIe bridge nodes
3. Correct the wifi@1,0 nodes to wifi@0,0 as explained in the commit itself

robimarko committed:
qualcommax: ipq807x, ipq60xx: backport DTS PCIe bridge node

Currently, we manually define PCI bridge nodes in devices that use
PCI cards, etc.

But since we will need to rework this for 6.12 anyway since upstream added
the PCIe bridge nodes[1][2] lets backport it now.

[1] https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/arch/arm64/boot/dts/qcom/ipq8074.dtsi?h=v6.12.28&id=ed3893f6f9b800ca774f63810c5f8838bc7cee78
[2] https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/arch/arm64/boot/dts/qcom/ipq6018.dtsi?h=v6.12.28&id=52358c64937e982d3cdcf64be58f08f30d8e518c

Link: https://github.com/openwrt/openwrt/pull/18789
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
qualcommax: remove PCIe bridge nodes

PCIe bridge nodes are now present in the SoC DTSI, so drop our downstream
ones as otherwise ath1*k BDF matching will fail.

Link: https://github.com/openwrt/openwrt/pull/18789
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko committed:
qualcommax: fix PCIe card wifi node address

Unlike other buses, for PCIe child nodes we should not use "wifi@1,0" as
the bus adress since for PCI devices the actual device adress on the bus is
actually set via the "reg" property and the node name is unused completely.

Though, it will cause DTC to throw warnings, so after investigation and
similar issue upstream[1] simply set the nodes to "wifi@0,0" instead.

[1] https://lists.infradead.org/pipermail/linux-arm-kernel/2021-August/680846.html

Link: https://github.com/openwrt/openwrt/pull/18789
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
@Ansuel, this should be the "proper" fix.

@georgemoussalem We might as well do IPQ50xx along the way by backporting you v9 PCI nodes patch if you agree?

robimarko commented:
ath10k and ath11k broke on 6.12 since we then had 2 bridge nodes since we still had our downstream bridge node so OF matching for properties got broke.

With this everything just works as before both on 6.6 and 6.12 on AX9000 which uses both ath11k and ath10k cards.

Here is lspci on AX9000:
```
root@OpenWrt:~# lspci -D -PP
0000:00:00.0 PCI bridge: Qualcomm Technologies, Inc Device 1002
0000:00:00.0/01:00.0 Network controller: Qualcomm Technologies, Inc QCN6024/9024/9074 Wireless Network Adapter (rev 01)
0001:00:00.0 PCI bridge: Qualcomm Technologies, Inc MDM9x55 LTE Modem [Snapdragon X16]
0001:00:00.0/01:00.0 Network controller: Qualcomm Atheros QCA9887 802.11ac Wireless Network Adapter
```

georgemoussalem commented:
@robimarko: I was planning to send v9 patches anyways, both the PHY and controller patches are in linix-next now, don’t know why the dts changes are taking so long (not just for pci).
Anyways, yes, let’s use the the entire v9 patches set and do the migration. While doing so, please also take note of the linux,pci-domain property change as that impacts the computed bus address too.

robimarko committed:
qualcommax: ipq50xx: use latest v9 PCIe DTS patch

Use the latest v9 PCIe DTS patch that is pending upstream, notable change
being that it includes PCIe bridge nodes.

Link: https://github.com/openwrt/openwrt/pull/18789
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
@georgemoussalem I will only use the latest DTS patch for this PR, PHY should be updated separately.
I already backported the upstreamed PCI controller patch after backporting the bulk clock commit as that made 6.12 way easier.

robimarko committed:
qualcommax: ipq50xx: add missing bus-range to PCIe bridges

It seems that bus-range property was missed for the PCIe bridges on
IPQ50xx, thus causing DTC to warn on any bus adress used on nodes
under the PCIe bridges.

Link: https://github.com/openwrt/openwrt/pull/18789
Signed-off-by: Robert Marko <robimarko@gmail.com>

georgemoussalem commented:
simply forgotten, I did check against dt_bindings_check and it didn't show up there.
Anyways, could you add it? I can send a new version of the patch upstream.

on the linux,pci-domain property, this will impact the hotplug script ath11k_caldata
for qcn9074:
`"ath11k/QCN9074/hw1.0/cal-pci-0001:01:00.0.bin")`
will need to change to 
`"ath11k/QCN9074/hw1.0/cal-pci-0000:01:00.0.bin")`

does it need a migration script or since it's in snapshot only we do without?

robimarko commented:
I have added it as a separate commit.

It would be great if we could migrate the path so it "just works TM"

robimarko committed:
qualcommax: ipq50xx: update PCI path for caldata

Now that IPQ50xx PCIe nodes have linux,pci-domain property defined
it changed the PCI path, but its now predictable and fixed.

So, update the caldata script accordingly.

Link: https://github.com/openwrt/openwrt/pull/18789
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
@georgemoussalem Can you post the wireless config of one of the boards that uses ath11k PCI cards on IPQ5018?
So I know what to migrate

georgemoussalem commented:
> @georgemoussalem Can you post the wireless config of one of the boards that uses ath11k PCI cards on IPQ5018? So I know what to migrate

```
config wifi-device 'radio1'
        option type 'mac80211'
        option path 'soc@0/a0000000.pcie/pci0001:00/0001:00:00.0/0001:01:00.0'
        option band '5g'
        option channel 'auto'
        option channels '36-64 100-144'
        option htmode 'HE160'
        option country 'US'
        option cell_density '0'
```

robimarko commented:
Great, off to lunch and I will whip up a script after that (Its balkan lunch so 2-3 hours)

robimarko committed:
qualcommax: ipq50xx: add PCI path migration script

PCI paths of IPQ50xx devices have changed now that linux,pci-domain is set
in the DTSI, so add a migration script for wireless config.

Link: https://github.com/openwrt/openwrt/pull/18789
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
@georgemoussalem I got held up, so whipped a quick script before heading off.

Please give it a go

georgemoussalem commented:
will test and I'll also prepare a separate PR to update the phy driver with the one accepted upstream

georgemoussalem commented on target/linux/qualcommax/ipq50xx/base-files/etc/hotplug.d/ieee80211/05-wifi-migrate:
```diff
@@ -0,0 +1,60 @@
+linksys,mx5500|\
+linksys,spnmx56)
+	migrate_radio 'soc@0/a0000000.pcie/pci0001:00/0001:00:00.0/0001:01:00.0' 'soc@0/a0000000.pcie/pci0000:00/0000:00:00.0/0000:01:00.0'
+    	;;
```
```
[   24.222325] ath11k_pci 0000:01:00.0: chip_id 0x0 chip_family 0x0 board_id 0xff soc_id 0xffffffff
[   24.222390] ath11k_pci 0000:01:00.0: fw_version 0x290b8862 fw_build_timestamp 2024-09-23 10:51 fw_build_id
[   24.529709] mtdblock: MTD device 'devinfo' is NAND, please consider using UBI block devices instead.
[   25.654294] ath11k_pci 0000:01:00.0: htt event 48 not handled
[   26.746231] mtdblock: MTD device 'rootfs' is NAND, please consider using UBI block devices instead.
```
migration script works as expected :+1: :

```
config wifi-device 'radio1'
        option type 'mac80211'
        option path 'soc@0/a0000000.pcie/pci0000:00/0000:00:00.0/0000:01:00.0'
        option band '5g'
        option channel 'auto'
        option channels '36-64 100-144'
        option htmode 'HE160'
        option country 'US'
        option cell_density '0'
```

robimarko replied:
Thanks, fixed the whitespace.

georgemoussalem replied:
Yes, it works. Didn’t need to enable wlan as I kept preserved settings and the migration script successfully updated the path in /etc/config/wireless

robimarko replied:
Yes, it updated it, but my worry is if the path is correct as I made "educated" guess on what it should be

georgemoussalem replied:
Well, wifi works, but so you want me to run any tests (lspci?)

robimarko replied:
If it works then thats it :)

testuser7 commented:
I just noticed that for AX3600 and AX9000 for the additional radio QCA9889 we use `nvmem-cell-names = "calibration"` and not `nvmem-cell-names = "pre-calibration"`:
- https://github.com/openwrt/openwrt/blob/main/target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8071-ax3600.dts#L66
- https://github.com/openwrt/openwrt/blob/main/target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-ax9000.dts#L578 

In this case should the BDF files be removed from the `firmware_qca-wireless` repository?
- https://github.com/openwrt/firmware_qca-wireless/blob/main/board-xiaomi_ax3600.qca9889
- https://github.com/openwrt/firmware_qca-wireless/blob/main/board-xiaomi_ax9000.qca9889

robimarko commented:
> I just noticed that for AX3600 and AX9000 for the additional radio QCA9889 we use `nvmem-cell-names = "calibration"` and not `nvmem-cell-names = "pre-calibration"`:
> 
> * https://github.com/openwrt/openwrt/blob/main/target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8071-ax3600.dts#L66
> * https://github.com/openwrt/openwrt/blob/main/target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8072-ax9000.dts#L578
> 
> In this case should the BDF files be removed from the `firmware_qca-wireless` repository?
> 
> * https://github.com/openwrt/firmware_qca-wireless/blob/main/board-xiaomi_ax3600.qca9889
> * https://github.com/openwrt/firmware_qca-wireless/blob/main/board-xiaomi_ax9000.qca9889

NO!

Wawe-1 ath10k cards use full calibration blobs while Wawe-2 use pre-calibration, but both require BDF-s.

testuser7 commented:
> Wawe-1 ath10k cards use full calibration blobs while Wawe-2 use pre-calibration, but both require BDF-s.

Are BDF files optional for QCA9889? In case of Meraki MR42/MR52 where the same radio chip is used calibration data is taken only from `ART` partition (there are no BDF files).

robimarko commented:
> > Wawe-1 ath10k cards use full calibration blobs while Wawe-2 use pre-calibration, but both require BDF-s.
> 
> Are BDF files optional for QCA9889? In case of Meraki MR42/MR52 where the same radio chip is used calibration data is taken only from `ART` partition (there are no BDF files).

AFAIK BDF-s are mandatory for ath10k onwards, some devices in OpenWrt did stupid stuff like symlink caldata as the BDF while others use upstream BDF-s
